pkg/util/kubernetes/apiserver/metadata_controller.go

```diff
@@ -46,18 +45,43 @@ type MetadataController struct {
 }
 
 // NewMetadataController returns a new metadata controller
-func NewMetadataController(nodeInformer coreinformers.NodeInformer, endpointsInformer coreinformers.EndpointsInformer) *MetadataController {
+func NewMetadataController(endpointsInformer coreinformers.EndpointsInformer, wmeta workloadmeta.Component) *MetadataController {
 	m := &MetadataController{
+		wmeta: wmeta,
 		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "endpoints"),
 	}
-	if _, err := nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    m.addNode,
-		DeleteFunc: m.deleteNode,
-	}); err != nil {
-		log.Errorf("error adding event handler to node informer: %f", err)
+
+	wmetaFilterParams := workloadmeta.FilterParams{
+		Kinds:     []workloadmeta.Kind{workloadmeta.KindKubernetesNode},
+		Source:    workloadmeta.SourceAll,
+		EventType: workloadmeta.EventTypeAll,
 	}
-	m.nodeLister = nodeInformer.Lister()
-	m.nodeListerSynced = nodeInformer.Informer().HasSynced
+
+	wmetaEventsCh := wmeta.Subscribe(
+		"metadata-controller",
+		workloadmeta.NormalPriority,
+		workloadmeta.NewFilter(&wmetaFilterParams),
+	)
```

Comment:
nit: I think if we declare these inside the go routine just below, we will be able to have `defer wmeta.Unsubscribe(...)` directly after `wmeta.Subscribe(...)`, which I think is a good practice.

Reply:
I added a new commit with this :+1: